app/controllers/RefusedAmountController.scala

```diff
@@ -39,52 +43,42 @@ class RefusedAmountController @Inject()(
                                          formProvider: RefusedAmountFormProvider,
                                          val controllerComponents: MessagesControllerComponents,
-                                         getCnCodeInformationService: GetCnCodeInformationService,
+                                         referenceDataService: ReferenceDataService,
                                          view: RefusedAmountView
                                        ) extends BaseNavigationController with AuthActionHelper {
 
   def onPageLoad(ern: String, arc: String, idx: Int, mode: Mode): Action[AnyContent] =
     authorisedDataRequestWithCachedMovementAsync(ern, arc) { implicit request =>
       withAddedItemAsync(idx) { item =>
-        getCnCodeInformationService.getCnCodeInformationWithMovementItems(Seq(item)).map {
-          serviceResult =>
-            val (item, cnCodeInformation) = serviceResult.head
-            val shortageAmount = request.userAnswers.get(ItemShortageOrExcessPage(item.itemUniqueReference)).flatMap(_.shortageAmount)
-            Ok(view(
-              fillForm(RefusedAmountPage(idx), formProvider(item.quantity, shortageAmount)),
-              routes.RefusedAmountController.onSubmit(ern, arc, idx, mode),
-              cnCodeInformation.unitOfMeasureCode.toUnitOfMeasure,
-              item,
-              cnCodeInformation,
-              idx
-            ))
-        }
+        renderView(Ok, fillForm(RefusedAmountPage(idx), form(item)), item, mode)
       }
     }
 
   def onSubmit(ern: String, arc: String, idx: Int, mode: Mode): Action[AnyContent] =
     authorisedDataRequestWithCachedMovementAsync(ern, arc) { implicit request =>
       withAddedItemAsync(idx) { item =>
-        val shortageAmount = request.userAnswers.get(ItemShortageOrExcessPage(item.itemUniqueReference)).flatMap(_.shortageAmount)
-        formProvider(item.quantity, shortageAmount).bindFromRequest().fold(
-          formWithErrors => {
-            getCnCodeInformationService.getCnCodeInformationWithMovementItems(Seq(item)).map {
-              serviceResult =>
-                val (_, cnCodeInformation) = serviceResult.head
-                BadRequest(view(
-                  formWithErrors,
-                  routes.RefusedAmountController.onSubmit(ern, arc, idx, mode),
-                  cnCodeInformation.unitOfMeasureCode.toUnitOfMeasure,
-                  item,
-                  cnCodeInformation,
-                  idx
-                ))
-            }
-          },
+        form(item).bindFromRequest().fold(
+          renderView(BadRequest, _, item, mode),
           value =>
             saveAndRedirect(RefusedAmountPage(idx), value, mode)
         )
       }
     }
-}
 
+  private def form(item: MovementItem)(implicit request: DataRequest[_]): Form[BigDecimal] = {
+    val shortageAmount = request.userAnswers.get(ItemShortageOrExcessPage(item.itemUniqueReference)).flatMap(_.shortageAmount)
```

Comment:
`.flatMap(_.shortageAmount)` isn't hit according to the coverage report - if it’d be a pain to add in then it can be done as a wider “get coverage up” story later though

Reply:
Will add a case